File: app/src/main/java/sk/romanstrazanec/particleflowkotlin/javacode/Particle.java
package sk.romanstrazanec.particleflowkotlin.javacode;

import android.graphics.Canvas;
import android.graphics.Paint;

class Particle {
    float x, y;
    private int color;

    Particle(float x, float y, int color) {
        this.x = x;
        this.y = y;
        this.color = color;
    }

    void moveTo(float x, float y) {
        this.x = x;
        this.y = y;
    }

    void draw(Canvas canvas, Paint paint) {
        paint.setColor(color);
        if (canvas != null)
            canvas.drawCircle(x, y, 1F, paint);
    }
}

File: app/src/main/java/sk/romanstrazanec/particleflowkotlin/Particle.kt
package sk.romanstrazanec.particleflowkotlin

import android.graphics.Canvas
import android.graphics.Paint

class Particle(var x: Float, var y: Float, private val r: Float, private val color: Int) {
    var prevX: Float = x
    var prevY: Float = y
    val dx get() = x - prevX
    val dy get() = y - prevY
    var dispersed: Boolean = false

    fun moveTo(x: Float, y: Float) {
        prevX = this.x.also { this.x = x }
        prevY = this.y.also { this.y = y }
    }

    fun draw(canvas: Canvas?, paint: Paint) {
        paint.color = color
        canvas?.drawCircle(x, y, r, paint)
        canvas?.drawLine(x, y, prevX, prevY, paint)
    }
}
